torch/csrc/jit/passes/batch_mm.cpp

```diff
@@ -0,0 +1,208 @@
+// procesed, and we can try to unify their TreeTokens (if they have them)
+// and build a larger tree.
+struct TreeToken {
+  uint64_t tree_size = 0; // NOTE: measured in number of leaves i.e. mm ops
+  std::array<int64_t, 2> lhs_sizes;
+  std::array<int64_t, 2> rhs_sizes;
+  Node *node = nullptr;
+  bool is_root = false;
+
+  static TreeToken fromMM(Node *mm) {
+    TreeToken token;
+    token.tree_size = 1;
+    Value *lhs = mm->inputs()[0];
+    Value *rhs = mm->inputs()[1];
+    token.lhs_sizes = as_array(lhs->type()->expect<TensorType>()->sizes());
+    token.rhs_sizes = as_array(rhs->type()->expect<TensorType>()->sizes());
+    token.node = mm;
+    token.is_root = true;
+    return token;
+  }
+
+  static TreeToken unify(Node *add, TreeToken& l, TreeToken& r) {
+    TreeToken token;
+    // See Note [Overlapping trees]
+    if (&l == &r || !l.is_root || !r.is_root)
+      return token;
+    // We can batch the tree only if all sizes match, because we need to
+    // cat inputs for both operands
+    if (l.lhs_sizes != r.lhs_sizes)
+      return token;
+    if (l.rhs_sizes != r.rhs_sizes)
+      return token;
+    token.tree_size = l.tree_size + r.tree_size;
+    token.lhs_sizes = l.lhs_sizes;
+    token.rhs_sizes = l.rhs_sizes;
+    token.node = add;
+    token.is_root = true;
+    l.is_root = r.is_root = false; // Reserve the subtrees, so they can't be used again.
+    return token;
+  }
+
+  operator bool() {
+    return is_root;
+  }
+
+  std::vector<Node*> gatherMatMuls() {
+    static Symbol mm_kind = "mm"_sym;
+    std::vector<Node*> matmuls;
+    std::vector<Node*> queue {node};
+    while (!queue.empty()) {
+      auto n = queue.back(); queue.pop_back();
+      if (n->kind() == mm_kind) {
+        matmuls.push_back(n);
+      } else {
+        queue.push_back(n->inputs()[0]->node());
+        queue.push_back(n->inputs()[1]->node());
+      }
+    }
+    return matmuls;
+  }
+};
+
+void BatchMM(std::shared_ptr<Graph>& graph) {
+  static Symbol mm_kind = "mm"_sym;
+  static Symbol add_kind = "add"_sym;
+  static Symbol cat_kind = "cat"_sym;
+  static Symbol dim_sym = "dim"_sym;
+
+  // Look for trees in the graph
+  std::unordered_map<Node*, TreeToken> tokens;
+  for (auto node : graph->nodes()) {
+    if (node->kind() == mm_kind) {
+      tokens[node] = TreeToken::fromMM(node);
+    } else if (node->kind() == add_kind) {
+      Node *lhs = node->inputs()[0]->node();
+      Node *rhs = node->inputs()[1]->node();
+      auto lhs_it = tokens.find(lhs);
+      auto rhs_it = tokens.find(rhs);
+      // See Note [Overlapping trees] (regarding the uses().size() == 1 check)
+      // We could treat a subtree with multiple uses as if it was overlapping.
+      if (lhs_it != tokens.end() && rhs_it != tokens.end() &&
+          lhs->output()->uses().size() == 1 && rhs->output()->uses().size() == 1) {
+        if (auto token = TreeToken::unify(node, lhs_it->second, rhs_it->second))
+          tokens[node] = token;
```

Comment:
It is a bit weird that tokens both have validity overheads, _and_ are only in the map if they are valid.  I was expecting one or the other.

Reply:
What do you mean by "validity overheads"? They are in the map even if they are invalid (they can become invalidated after insertion and are never removed). This line is just an optimization to stop expanding the trees when it's unnecessary. It would be safe to insert them unconditionally.

Reply:
I meant overload rather than overhead. I just read the code wrong and didn't realize is_root was the boolean condition being checked using the overloaded cast operator. I thought validity meant whether or not it was is any tree but not necessarily a root, and is_root was checked separately. As it actual is, I think it makes sense.